### PR TITLE
Serial console

### PR DESCRIPTION
**Pull request by jreidinger**
PBI: https://trello.com/c/c66dIsRy/1417-3-15-ga-please-dont-throw-an-exception-on-valid-grubterminal-options

how it looks:
bootloader_qt_terminal
bootloader_ncurses_terminal

### CODE REVIEW

File: src/modules/Bootloader.rb [hunk 1]
```diff
@@ -146,10 +146,17 @@ def Read
         ::Bootloader::BootloaderFactory.current = ::Bootloader::BootloaderFactory.proposed
         ::Bootloader::BootloaderFactory.current.propose
       rescue ::Bootloader::BrokenConfiguration => e
-        ret = Yast::Report.AnyQuestion(_("Broken Configuration"),
+        msg = if e.is_a?(::Bootloader::BrokenConfiguration)
```
Comment (lslezak) [inline src/modules/Bootloader.rb]: This looks strange with the `rescue ::Bootloader::BrokenConfiguration` statement above, `e` cannot be something else. Or is `::Bootloader::UnsupportedOption` missing in the `rescue`?

File: src/lib/bootloader/grub2base.rb [hunk 2]
```diff
@@ -217,11 +212,15 @@ def merge_grub_default(other)
...
-        val = other.public_send(attr)
-        default.public_send((attr.to_s + "=").to_sym, val) if val
+      begin
+        # string attributes
+        [:serial_console, :terminal, :timeout, :hidden_timeout, :distributor,
+         :gfxmode, :theme, :default].each do |attr|
+          val = other.public_send(attr)
+          default.public_send((attr.to_s + "=").to_sym, val) if val
+        end
+      rescue RuntimeError
```
Comment (lslezak) [inline src/lib/bootloader/grub2base.rb]: Um, I'm not sure if catching all run time exceptions is a good approach here. Maybe a comment would help here...

File: src/lib/bootloader/exceptions.rb [hunk 3]
```diff
@@ -26,6 +26,23 @@ def initialize(msg)
...
+    include Yast::I18n
+    attr_reader :option
+
+    def initialize(option)
+      @reason = option
+      textdomain "bootloader"
+
+      # TRANSLATORS: %s is name of option that has unsupported value
+      super _("Option '%s' contain unsupported value.\nYaST2 Bootloader cannot handle it.\n" \
+        "To use YaST2 Bootloader configuration have to be reproposed\n" \
```
Comment (lslezak) [inline src/lib/bootloader/exceptions.rb]: IMHO this part is too complicated, simpler

`The configuration needs to be proposed again or edited manually.` 

would be better.

File: src/lib/bootloader/exceptions.rb [hunk 4]
```diff
@@ -26,6 +26,23 @@ def initialize(msg)
...
+  class UnsupportedOption < RuntimeError
+    include Yast::I18n
+    attr_reader :option
+
+    def initialize(option)
+      @reason = option
+      textdomain "bootloader"
+
+      # TRANSLATORS: %s is name of option that has unsupported value
+      super _("Option '%s' contain unsupported value.\nYaST2 Bootloader cannot handle it.\n" \
```
Comment (lslezak) [inline src/lib/bootloader/exceptions.rb]: "contains an unsupported value"

File: src/lib/bootloader/config_dialog.rb [hunk 5]
```diff
@@ -19,11 +19,17 @@ class ConfigDialog
...
-          "Propose new configuration from scratch?") % e.reason,
+          "Propose new configuration from scratch?") % e.reason
+      else
+        e.message
+      end
+
+      ret = Yast::Report.AnyQuestion(_("Unsupported Configuration"),
+        msg,
         _("Propose"),
         _("Quit"),
```
Comment (lslezak) [inline src/lib/bootloader/config_dialog.rb]: "Quit" means exiting which does not make much sense during installation. The message says "repropose or edit manually" so the appropriate label could be `Edit` or something like that.

File: src/modules/Bootloader.rb [hunk 6]
```diff
@@ -146,10 +146,17 @@ def Read
...
+            "Propose new configuration from scratch?") % e.reason
+        else
+          e.message
+        end
+
+        ret = Yast::Report.AnyQuestion(_("Unsupported Configuration"),
+          # TRANSLATORS: %s stands for readon why yast cannot process it
+          msg,
           _("Propose"),
           _("Quit"),
```
Comment (lslezak) [inline src/modules/Bootloader.rb]: Is `Quit` also OK here?

File: src/modules/Bootloader.rb (hunk 1 above)
Comment (jreidinger) [reply src/modules/Bootloader.rb]: yes, it is missing.

File: src/lib/bootloader/grub2base.rb (hunk 2 above)
Comment (jreidinger) [reply src/lib/bootloader/grub2base.rb]: yep, but cfa_grub2 raise generic RuntimeError, I hope it is just temporary solution. Also cfa_grub2 can be improved to raise specific exception, but currently for SLE15 GA I do not want to touch it.

File: src/lib/bootloader/config_dialog.rb (hunk 5 above)
Comment (jreidinger) [reply src/lib/bootloader/config_dialog.rb]: IT cannot happen during installation. It is runtime only where user edit configuration with something else then YaST2.

File: src/modules/Bootloader.rb (hunk 6 above)
Comment (jreidinger) [reply src/modules/Bootloader.rb]: yes, Quit exits YaST2.

File: src/lib/bootloader/grub2base.rb (hunk 2 above)
Comment (lslezak) [reply src/lib/bootloader/grub2base.rb]: OK, then just add comment with that info.

File: src/lib/bootloader/config_dialog.rb (hunk 5 above)
Comment (jreidinger) [reply src/lib/bootloader/config_dialog.rb]: And edit manually means edit by something other than yast2 bootloader.

**Pull request merged**